### PR TITLE
Add producer batch send queue size limit

### PR DESCRIPTION
**Pull request by toidi**
As noted in #297 unlimited queue size can cause memory leak.

My patches add ability to specify queue size.
By default queue size is unlimited.

If queue become full next send message will raise an `BatchQueueOverfilledError`.

### CODE REVIEW

File: kafka/producer/base.py [hunk 1]
```diff
@@ -139,7 +143,8 @@ def __init__(self, client, async=False,
             log.warning("async producer does not guarantee message delivery!")
             log.warning("Current implementation does not retry Failed messages")
             log.warning("Use at your own risk! (or help improve with a PR!)")
-            self.queue = Queue()  # Messages are sent through this queue
+            # Messages are sent through this queue
+            self.queue = Queue(maxsize=batch_send_queue_maxsize)
```
Comment (dpkp) [inline kafka/producer/base.py]: this is used both in batch mode as well as in simple async mode -- the variable name probably should not be specific to batch mode

File: test/test_producer.py [hunk 2]
```diff
@@ -25,3 +26,16 @@ def test_producer_message_types(self):
             # This should not raise an exception
             producer.send_messages(topic, partition, m)
 
+    @patch('kafka.producer.base.Process')
+    def test_producer_batch_send_queue_overfilled(self, process_mock):
```
Comment (dpkp) [inline test/test_producer.py]: should also test `async=True` mode

**Comment by toidi** [comment]
@dpkp how producer should behave with async mode enabled?

**Comment by dpkp** [comment]
should be the same behavior, i just think that the max queue size variable should probably be labeled as an `async_` kwarg and applied in the same way whether async mode is enabled directly via `async=True` or indirectly via `batch_send=True`

File: kafka/common.py [hunk 3]
```diff
@@ -200,6 +200,10 @@ class KafkaConfigurationError(KafkaError):
     pass
 
 
+class BatchQueueOverfilledError(KafkaError):
```
Comment (dpkp) [inline kafka/common.py]: How about `AsyncQueueFull`

File: kafka/producer/base.py [hunk 4]
```diff
@@ -188,7 +194,13 @@ def _send_messages(self, topic, partition, *msg, **kwargs):
 
         if self.async:
             for m in msg:
-                self.queue.put((TopicAndPartition(topic, partition), m, key))
+                try:
+                    item = (TopicAndPartition(topic, partition), m, key)
+                    self.queue.put_nowait(item)
+                except Full:
+                    raise BatchQueueOverfilledError(
+                        'Producer batch send queue overfilled. '
```
Comment (dpkp) [inline kafka/producer/base.py]: s/batch/async/

**Comment by dpkp** [comment]
consider whether PR 283 is a more robust approach (includes backoff + retry)

**Comment by rogaha** [comment]
+1

**Comment by rogaha** [comment]
What is the status of this PR? Is there anything I can help you guys with in order to get this merged?

**Comment by rogaha** [comment]
ping @dpkp

**Comment by toli** [comment]
What's the status of this PR? it's been 4+ months, same with https://github.com/mumrah/kafka-python/pull/283
We are facing a similar problem at at work, and i'd prefer not to have custom patches and just use vanilla upstream code...

@dpkp - you mentioned something about Rdio having their own, perhaps you can get that committed to upstream? or get this patch in?

**Comment by dpkp** [comment]
I'd like to get #331 merged for this issue instead. If you have a second to review, would appreciate more eyeballs and/or testing.

**Comment by dpkp** [comment]
#331 is merged and supports this via `async_queue_maxsize` and `async_queue_put_timeout` configuration parameters to the producer

**Pull request closed without merging**